Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractNodeStore.java

```diff
@@ -0,0 +1,527 @@
+        cutValue = new float[capacity - 1];
+        leafMass = new HashMap<>();
+        int cache_limit = (int) Math.floor(nodeCacheFraction * capacity);
+        rangeSumData = new double[cache_limit];
+        boundingBoxData = new float[2 * dimensions * cache_limit];
+    }
+
+    abstract int addNode(int parentIndex, int leftIndex, int rightIndex, int cutDimension, double cutValue, int mass);
+
+    public int addLeaf(int parentIndex, int pointIndex, int mass) {
+        return pointIndex + capacity;
+    }
+
+    public boolean isLeaf(int index) {
+        return index != 0 && index >= capacity;
+    }
+
+    public abstract int getLeftIndex(int index);
+
+    public abstract int getRightIndex(int index);
+
+    public abstract void setRoot(int index);
+
+    public void increaseLeafMass(int index) {
+        int y = (index - capacity);
+        leafMass.merge(y, 1, Integer::sum);
+    }
+
+    public int decreaseLeafMass(int index) {
+        int y = (index - capacity);
+        Integer value = leafMass.remove(y);
+        if (value != null) {
+            if (value > 1) {
+                leafMass.put(y, (value - 1));
+                return value;
+            } else {
+                return 1;
+            }
+        } else {
+            return 0;
+        }
+    }
+
+    public int translate(int index) {
+        if (index != 0 && rangeSumData.length <= index - 1) {
+            return Integer.MAX_VALUE;
+        } else {
+            return index - 1;
+        }
+    }
+
+    public void copyBoxToData(int index, BoundingBoxFloat box) {
+        int idx = translate(index);
+        if (idx != Integer.MAX_VALUE) {
+            int base = 2 * idx * dimensions;
+            int mid = base + dimensions;
+            System.arraycopy(box.getMinValues(), 0, boundingBoxData, base, dimensions);
+            System.arraycopy(box.getMaxValues(), 0, boundingBoxData, mid, dimensions);
+            rangeSumData[idx] = box.getRangeSum();
+        }
+    }
+
+    public boolean checkContainsAndAddPoint(int index, float[] point) {
+        int idx = translate(index);
+        if (idx != Integer.MAX_VALUE) {
+            int base = 2 * idx * dimensions;
+            int mid = base + dimensions;
+            double rangeSum = 0;
+            for (int i = 0; i < dimensions; i++) {
+                boundingBoxData[base + i] = Math.min(boundingBoxData[base + i], point[i]);
+            }
+            for (int i = 0; i < dimensions; i++) {
+                boundingBoxData[mid + i] = Math.max(boundingBoxData[mid + i], point[i]);
+            }
+            for (int i = 0; i < dimensions; i++) {
+                rangeSum += boundingBoxData[mid + i] - boundingBoxData[base + i];
+            }
+            boolean answer = (rangeSumData[idx] == rangeSum);
+            rangeSumData[idx] = rangeSum;
+            return answer;
+        }
+        return false;
+    }
+
+    public BoundingBoxFloat getBox(int index, IPointStoreView<float[]> pointStoreView) {
```

Comment:
It seems strange to me to make the point store view a method parameter and not an instance field. By having it as a method argument, that suggests that you should be able to pass any point store view in and get some kind of answer.

Reply:
I think the 2.0 version called for a more lightweight nodeview -- let me revisit this, it's possible that we can simplify these.